# Numpy

Pada topik ini kita akan membahas sebuah library yang bisa dimanfaatkan untuk mengolah data dalam bentuk array. Library tersebut bernama ``Numpy``. Dokumentasi lengkap perihal numpy dapat dilihat pada https://numpy.org/.

Sebelum bisa menggunakan module ``numpy``, pastikan module ini sudah dipasang di lingkungan Python yang digunakan. Perintah instalasinya sebagai berikut:

<b>Instalasi via pip</b> <br>
``pip3 install numpy``

<b>Instalasi via conda</b> (secara default sudah tersedia ketika instalasi Anaconda atau Miniconda) <br>
``conda install numpy``

Setelah melakukan instalasi, mari kita masuk ke pembahasan module ``numpy`` ini.

## Numpy Array vs. Python List

Numpy memiliki struktur data yang mirip seperti List, yaitu numpy array. Numpy array lebih efisien dalam penggunaan resource dan lebih cepat dalam pengaksesannya. Berikut adalah contoh penggunaan numpy array:

In [54]:
import numpy as np

array = np.array([1, 'a', 2, 'b'])
print(array)
print(array[1])
print(array[1:3])
print(array[:3])
print(array[1:])
print(array[-1])
print(array[-1:-3:-1])

['1' 'a' '2' 'b']
a
['a' '2']
['1' 'a' '2']
['a' '2' 'b']
b
['b' '2']


Tampilan array-nya sama dengan list yang ada pada python bahkan operasi slicingnya sama dengan list yang ada di Python. Jadi tidak perlu khawatir untuk menggunakan array dari numpy ini. Kita juga bisa mengkonversi list python menjadi numpy array dengan cara type casting. Berikut cara penggunaannya:

In [55]:
import numpy as np

list_python = [1, 'a', 2, 'b', 'adam']

array_numpy = np.array(list_python)
print(type(array_numpy))

<class 'numpy.ndarray'>


Sama seperti list, numpy array juga bisa direpresentasikan menjadi array dua dimensi. Berikut adalah contoh penggunaanya:

In [56]:
import numpy as np

list_2d = [[10, 20, 30],[40, 50, 60], [70, 80, 90]]
print(list_2d)

print()

array_2d = np.array(list_2d)
print(array_2d)

[[10, 20, 30], [40, 50, 60], [70, 80, 90]]

[[10 20 30]
 [40 50 60]
 [70 80 90]]


Kita bisa melihat bahwa numpy array lebih representatif untuk kita gunakan.

### Generate Array di Numpy

Numpy memiliki konsep generator yang mirip seperti list di Python. Apabila di list kita menggunakan ``range()`` yang disimpan dalam sebuah struktur perulangan, maka kita bisa menggunakan method ``arange()`` di numpy. Secara konsep, fungsi ini sebagai pengganti generator ``range()``. Berikut contoh penggunaannya:

In [3]:
import numpy as np

# List di Python
list_bil = []
for i in range(0,11):
    list_bil.append(i)

print(list_bil)

# numpy array
array = np.arange(0,11)
print(array)

array = np.arange(0,11,2)
print(array)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[ 0  1  2  3  4  5  6  7  8  9 10]
[ 0  2  4  6  8 10]


Selain untuk meng-generate array seperti contoh tadi, numpy memiliki kemampuan untuk membuat array berisi bilangan 0 (berguna untuk operasi matriks). Fungsi yang digunakan adalah ``zeros()``. Berikut cara penggunaannya:

In [58]:
import numpy as np

array_nol = np.zeros(10)
print(array_nol)

print()

array_nol_2d = np.zeros((10, 10))
print(array_nol_2d)

print()

array_nol_2d = np.zeros((10, 10), dtype= int)
print(array_nol_2d)


[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]

[[0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0 0]]


Selain ``zeros()`` numpy juga memiliki method ``ones()`` yang fungsinya mirip seperti ``zeros()`` akan tetapi nilai ynag di-generate adalah 1.

In [59]:
import numpy as np

array_satu = np.ones(10)
print(array_satu)

print()

array_satu_2d = np.ones((5, 10))
print(array_satu_2d)

print()

array_satu_2d = np.ones((5, 10),dtype=int)
print(array_satu_2d)


[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

[[1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1 1 1]]


### Generate Spaced Number di Numpy

Kita bisa menggunakan fungsi ``linspace()`` untuk generate spaced number pada interval yang ditentukan. Berikut contoh penggunaannya:

In [60]:
import numpy as np

matriks = np.linspace(0,10)
print(matriks)

[ 0.          0.20408163  0.40816327  0.6122449   0.81632653  1.02040816
  1.2244898   1.42857143  1.63265306  1.83673469  2.04081633  2.24489796
  2.44897959  2.65306122  2.85714286  3.06122449  3.26530612  3.46938776
  3.67346939  3.87755102  4.08163265  4.28571429  4.48979592  4.69387755
  4.89795918  5.10204082  5.30612245  5.51020408  5.71428571  5.91836735
  6.12244898  6.32653061  6.53061224  6.73469388  6.93877551  7.14285714
  7.34693878  7.55102041  7.75510204  7.95918367  8.16326531  8.36734694
  8.57142857  8.7755102   8.97959184  9.18367347  9.3877551   9.59183673
  9.79591837 10.        ]


In [61]:
import numpy as np

matriks = np.linspace(0, 10, 100)
print(matriks)

[ 0.          0.1010101   0.2020202   0.3030303   0.4040404   0.50505051
  0.60606061  0.70707071  0.80808081  0.90909091  1.01010101  1.11111111
  1.21212121  1.31313131  1.41414141  1.51515152  1.61616162  1.71717172
  1.81818182  1.91919192  2.02020202  2.12121212  2.22222222  2.32323232
  2.42424242  2.52525253  2.62626263  2.72727273  2.82828283  2.92929293
  3.03030303  3.13131313  3.23232323  3.33333333  3.43434343  3.53535354
  3.63636364  3.73737374  3.83838384  3.93939394  4.04040404  4.14141414
  4.24242424  4.34343434  4.44444444  4.54545455  4.64646465  4.74747475
  4.84848485  4.94949495  5.05050505  5.15151515  5.25252525  5.35353535
  5.45454545  5.55555556  5.65656566  5.75757576  5.85858586  5.95959596
  6.06060606  6.16161616  6.26262626  6.36363636  6.46464646  6.56565657
  6.66666667  6.76767677  6.86868687  6.96969697  7.07070707  7.17171717
  7.27272727  7.37373737  7.47474747  7.57575758  7.67676768  7.77777778
  7.87878788  7.97979798  8.08080808  8.18181818  8

### Matriks Indentitas

Numpy memiliki fungsi ``eye()`` yang memungkinkan programmer untuk membuat matriks identitas dengan jumlah tertentu. Pengertian dan kegunaan matriks identitas dapat dipelajari singkat di https://id.wikipedia.org/wiki/Matriks_identitas. Berikut contoh penggunaannya:

In [62]:
import numpy as np

matriks_identitas = np.eye(5)
print(matriks_identitas)

print()

matriks_identitas = np.eye(3)
print(matriks_identitas)


[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


### Generate Array Random

Numpy memiliki fungsi yang memungkinkan programmer membuat array berisi data random. Konsep random di numpy memiliki mekanisme yang mirip dengan Python. Berikut contoh penggunaannya:

In [63]:
import numpy as np

rand_array = np.random.rand(5)
print(rand_array)

print()

rand_matriks = np.random.rand(5,5)
print(rand_matriks)

[0.52225942 0.77136632 0.20063995 0.98574679 0.14381081]

[[0.48237973 0.16617865 0.40156018 0.63494413 0.70562708]
 [0.29306454 0.51360333 0.1723148  0.30917908 0.44439935]
 [0.9256396  0.5725619  0.90901124 0.8866892  0.57293701]
 [0.67351231 0.2651699  0.76102096 0.40849587 0.90177462]
 [0.9071997  0.65953779 0.86025361 0.34868705 0.35453378]]


Selain fungsi tersebut, numpy bisa menghasilkan random array yang berisi angka random untuk distribusi normal standar. Berikut contoh penggunaannya:

In [64]:
import numpy as np

rand_normal_array = np.random.randn(5)
print(rand_normal_array)

print()

rand_normal_matriks = np.random.randn(4,4)
print(rand_normal_matriks)

[-0.74033885  1.25878343  1.42331506  0.36543043  2.66486957]

[[-0.94526199 -1.90062742 -0.45981376 -0.8306381 ]
 [-0.85583357 -0.1847662   1.42916689 -2.02005298]
 [-0.32324644  1.29395179  1.23669254  0.86259919]
 [-1.19288166 -1.93608011 -0.76375318  1.02649718]]


Fungsi random dari numpy yang akan dibahas terakhir adalah ``randint()``. Fungsi ini memiliki kemiripan dengan fungsi ``randint()`` di python. Berikut contoh penggunaannya:

In [65]:
import numpy as np

nilai_random = np.random.randint(1,10)
print(nilai_random)

rand_int_array = np.random.randint(1,10,10)
print(rand_int_array)

4
[2 8 1 2 6 9 5 4 5 4]


### Fungsi reshape()

Fungsi ``reshape()`` memungkinkan kita untuk membentuk suatu array dengan dimensi menjadi array dengan dimensi lain. Berikut contoh penggunaannya:

In [66]:
import numpy as np

array = np.random.randint(1,10,25)
print(array)

print()

matriks_55 = array.reshape(5,5)
print(matriks_55)

[4 1 9 4 4 6 3 8 6 7 5 8 4 5 4 2 7 1 8 7 5 3 4 1 8]

[[4 1 9 4 4]
 [6 3 8 6 7]
 [5 8 4 5 4]
 [2 7 1 8 7]
 [5 3 4 1 8]]


><b>Catatan:</b> Jumlah elemen pada array yang mau dibentuk ulang harus sama dengan jumlah elemen pada dimensi array yang baru.

## Operasi pada Array

Numpy bisa kita gunakan untuk mempermudah perhitungan atau operasi-operasi pada array atau matriks. Berikut adalah bahasan tentang operasi pada array dan matriks yang bisa didukung (sering digunakan) oleh numpy.

### Pertambahan Array

Layaknya di ilmu Matematika, array bisa ditambahkan dengan array lainnya. Berikut contoh operasi pertambahan pada array menggunakan numpy:

In [67]:
import numpy as np

array1 = np.array([1, 2, 3])
array2 = np.ones(3, dtype = int)

array_hasil = array1 + array2

print(array_hasil)

print()

array1 = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]])
array2 = np.ones((1, 3), dtype=int)

array_hasil = array1 + array2

print(array_hasil)

[2 3 4]

[[2 3 4]
 [2 3 4]
 [2 3 4]]


### Pengurangan Array

Operasi aritmatika berikutnya yang didukung adalah operasi pengurangan. Berikut adalah contoh penggunaanya:

In [68]:
import numpy as np

array1 = np.array([1, 2, 3])
array2 = np.ones(3, dtype=int)

array_hasil = array1 - array2

print(array_hasil)

print()

array1 = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]])
array2 = np.ones((1, 3), dtype=int)

array_hasil = array1 - array2

print(array_hasil)

[0 1 2]

[[0 1 2]
 [0 1 2]
 [0 1 2]]


### Perkalian Array

Berikutnya adalah operasi perkalian yang bisa dilihat pada contoh kode program berikut:

In [69]:
import numpy as np

array1 = np.array([1, 2, 3])
array2 = np.ones(3, dtype=int)

array_hasil = array1 * array2

print(array_hasil)

print()

array1 = array2 = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]])

array_hasil = array1 * array2

print(array_hasil)

[1 2 3]

[[1 4 9]
 [1 4 9]
 [1 4 9]]


><b>Catatan:</b> Tetap perhatikan aturan perkalian pada matriks (dimensi harus memenuhi aturan mxn * n*o)

### Pembagian Array

Mirip seperti perkalian, berikut contoh dari operasi pembagian pada array:

In [70]:
import numpy as np

array1 = np.array([1, 2, 3])
array2 = np.ones(3, dtype=int)

array_hasil = array2 / array1

print(array_hasil)

print()

array1 = array2 = np.array([[1, 2, 3], [1, 2, 3], [1, 2, 3]])

array_hasil = array1 / array2

print(array_hasil)


[1.         0.5        0.33333333]

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


### Agregat Fungsi pada Numpy

Array pada numpy bisa dikenakan operasi agregat. Tentunya hal ini sangat berguna dalam melakukan pengolahan data. Operasi agregat yang bisa dikenakan pada array numpy dapat dilihat pada tabel berikut:

| Operasi Agregat | Penjelasan |
| max() | Mengembalikan nilai berupa elemen terbesar di array |
| min() | Mengembalikan nilai berupa elemen terkecil di array |
| sum() | Menjumlahkan seluruh elemen pada array |

Berikut contoh penggunaannya:

In [71]:
import numpy as np

array = np.array([1, 2, 3, 30, 21, 15, 110, 101, 102])

print(array.min())
print(array.max())
print(array.sum())

print()

array = np.array([[1, 2, 3], [30, 21, 15], [110, 101, 102]])

print(array.min())
print(array.max())
print(array.sum())


1
110
385

1
110
385


### Unique value di Array

Apabila kita membutuhkan array yang isinya unique dan juga membutuhkan berapa jumlah elemen unique-nya, kita bisa melakukan mekanisme berikut:

In [72]:
import numpy as np

array = np.array([11, 11, 12, 13, 14, 15, 16, 17, 12, 13, 11, 14, 18, 19, 20])
print(array)

array_unique = np.unique(array)
print(array_unique)


[11 11 12 13 14 15 16 17 12 13 11 14 18 19 20]
[11 12 13 14 15 16 17 18 19 20]


Kita juga bisa mengetahui posisi/letak dari nilai yang menyatakan indeks pertama dari nilai unique tersebut, kita bisa menggunakan parameter ``return_index = True`` di fungsi ``unique()`` sebagai berikut:

In [73]:
import numpy as np

array = np.array([11, 11, 12, 13, 14, 15, 16, 17, 12, 13, 11, 14, 18, 19, 20])
print(array)

array_unique, index_unique = np.unique(array, return_index = True)
print(array_unique)
print(index_unique)

[11 11 12 13 14 15 16 17 12 13 11 14 18 19 20]
[11 12 13 14 15 16 17 18 19 20]
[ 0  2  3  4  5  6  7 12 13 14]


Selain parameter ``return_index``, kita bisa menambahkan parameter ``return_counts`` untuk mendapatkan jumlah elemen unique pada array. Contohnya sebagai berikut:

In [74]:
import numpy as np

array = np.array([11, 11, 12, 13, 14, 15, 16, 17, 12, 13, 11, 14, 18, 19, 20])
print(array)

array_unique, jml_elemen = np.unique(array, return_counts = True)
print(array_unique)
print(jml_elemen)


[11 11 12 13 14 15 16 17 12 13 11 14 18 19 20]
[11 12 13 14 15 16 17 18 19 20]
[3 2 2 2 1 1 1 1 1 1]


### Transpose Matriks

Kita bisa menggunakan fungsi ``reshape()`` pada matriks untuk membantu melakukan operasi transpose matriks. Gambaran operasi transpose pada matriks dapat dilihat pada gambar berikut:

<img src = "https://idschool.net/wp-content/uploads/2017/11/Transpose-Matriks-1-e1510690764387.png" alt = "Ilustrasi Transpose Matriks">

Berikut contoh kasusnya:

In [75]:
import numpy as np

array = np.array([[1, 2, 3], [4, 5, 6]])
print(array)

print()

array_transpose = array.reshape(3,2)
print(array_transpose)

[[1 2 3]
 [4 5 6]]

[[1 2]
 [3 4]
 [5 6]]


### Reverse Array

Kita bisa menggunakan fungsi ``flip()`` untuk me-reverse elemen pada array. Berikut contoh penggunaannya:

In [76]:
import numpy as np

array = np.array([1, 2, 3, 4, 5, 6, 7, 8])
print(array)

print()

array_reverse = np.flip(array)
print(array_reverse)

[1 2 3 4 5 6 7 8]

[8 7 6 5 4 3 2 1]


In [77]:
import numpy as np

array = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
print(array)

print()

array_reverse = np.flip(array)
print(array_reverse)


[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

[[12 11 10  9]
 [ 8  7  6  5]
 [ 4  3  2  1]]


### Flattening Matrik ke Array

Kita bisa membuat matriks (array dua dimensi) menjadi array 1 dimensi dengan menggunakan fungsi ``flatten()``. Berikut contohnya:

In [80]:
import numpy as np

matriks = np.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
print(matriks)

print()

array = matriks.flatten()
print(array)


[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]

[ 1  2  3  4  5  6  7  8  9 10 11 12]


### Hands on Lab 1: Bermain dengan kecepatan

Diberikan data jarak dan waktu tempuh sebagai berikut:

| Jarak (Km) | Waktu Tempuh (Jam) |
| :--- | :--- |
| 100 | 2.5 |
| 84 | 2.3 |
| 200 | 3.4 |
| 5 | 0.1 |

Carilah:

1. Kecepatan untuk masing-masing perjalanan
2. Kecepatan rata-rata dari seluruh perjalanan
3. Kecepatan paling cepat dan paling lambat

### Hands on Lab 2: Bermain dengan nilai

Diberikan data nilai akhir dari kelas IPA 1 (100 siswa) untuk mata kuliah Biologi sebagai berikut:

``75, 70, 65, 50, 73, 58, 53, 69, 77, 62, 61, 70, 54, 66, 53, 51, 64,
52, 50, 77, 68, 63, 61, 68, 58, 67, 63, 67, 72, 70, 71, 60, 68, 75,
74, 62, 61, 76, 52, 54, 73, 72, 66, 56, 76, 60, 54, 51, 70, 70, 51,
67, 57, 67, 74, 56, 61, 54, 58, 64, 60, 54, 67, 74, 61, 72, 57, 54,
70, 55, 62, 61, 52, 60, 64, 64, 56, 79, 75, 73, 65, 64, 68, 69, 63,
76, 76, 76, 56, 51, 61, 70, 79, 55, 66, 66, 58, 53, 52, 57``

Hitung:
1. Rata-rata nilai
2. Hitung berapa kemunculan dari setiap nilai yang ada
3. Nilai terbesar dan terkecil
4. Bentuk kelompok nilai. Setiap kelompok berisi 10 nilai siswa. Hitung poin 1 sampai 3 untuk setiap kelompoknya.